Cache actions and events queries between event processings

fuziontech opened the pull request:
## Changes
Cache sql and params between event processing. Building the query takes some cpu time so this speeds things up to about 70ms per event from 180ms per event before on my laptop.

## Checklist
- [x] All querysets/queries filter by Team (if applicable)
- [x] Backend tests (if applicable)
- [x] Cypress E2E tests (if applicable)

fuziontech committed:
Cache actions and events queries between event processings

timgl commented on posthog/models/event.py:
```diff
@@ -237,35 +246,58 @@ def person(self):
-                }
-            )
-        event = [event for event in events][0]
+        if not self.team_id in LAST_UPDATED_TEAM_ACTION \
```
Maybe personal preference but I think these two if statements could be split out to different functions

fuziontech replied:
I think I know what you are saying here. I'll update PR and see if we are on same page

timgl commented on posthog/models/event.py:
```diff
@@ -237,35 +246,58 @@ def person(self):
+                )
+            q, p = events.query.sql_with_params()
+            qp = tuple(['%s' if i == self.pk else i for i in p])
+            qcache = {self.event: (q, qp)}
```
These 2 lines are a bit magic (especially with the non-descriptive variable names). Is there a way of doing this that's slightly less magic or just more descriptive?

fuziontech replied:
Totally agree. I'll work that in too.

timgl commented on posthog/models/event.py:
```diff
@@ -237,35 +246,58 @@ def person(self):
+            or not self.team_id in TEAM_EVENT_ACTION_QUERY_CACHE \
+            or not self.event in TEAM_EVENT_ACTION_QUERY_CACHE[self.team_id] \
+            or not self.team_id in TEAM_ACTION_QUERY_CACHE \
+            or (last_updated_action_ts > LAST_UPDATED_TEAM_ACTION[self.team_id]):
```
Is this logic implicitly tested by the current tests, or should we add another test to test this logic specifically? I think we do have some tests that test updating of actions, just want to make sure they would hit this. Maybe a specific test would be good?

fuziontech replied:
I think a specific test would be the right way to do this. That's what I was going to hack on a little later if I have time.

fuziontech committed:
feedbacked

timgl commented:
@fuziontech Sorry I missed this. Looks like the rebase went a bit awry. once that's done I'll qa and merge :)

fuziontech commented:
Woops!

fuziontech committed:
BaseTest -> TransactionBaseTest, BaseTest is causing problems

fuziontech commented:
Cleaned this up a bit. I had a really hard time figuring out why tests broke after rebasing. After quite a bit of debugging I changed the `BaseTest` from `TestCase` to [TransactionTestCase](https://docs.djangoproject.com/en/3.0/topics/testing/tools/#transactiontestcase) which is the parent class for BaseTest and forces the DB to commit changes before moving on instead of rolling back each Test in a transaction. The downside for this is that after each test under this parent class all tables are truncated. This slows down testing significantly. I happened upon this because I was running out of leads and enabled the `TransactionTestCase` just so that I could inspect the data in the test DB.

TODO for me is to research and understand why these tests fail under `TestCase` and succeed under `TransactionTestCase`.
